_posts/2015-10-30-october-newsletter.md

```diff
@@ -0,0 +1,292 @@
+    - Is based on msgpack and explicitly standardized in documentation which
+      allows creation of plugins/tools that perform arbitrary manipulations.
+    - Assumes no state stored between data pieces inside the file, which makes
+      such tools simpler and allows such manipulations as “to concatenate two
+      ShaDa files from different NeoVim instances simply use `cat 1.shada
+      2.shada > joined.shada`, NeoVim will handle this properly when reading”.
+    - Uses expanded paths(/home instead of ~/).
+- Embedded timestamps that allow multiple NeoVim instances to correctly merge
+  their data when writing or reading ShaDa files.
+- Forward compatibility includes NeoVim core code that preserves additional
+  information found in ShaDa files generated by newer NeoVim versions in case
+  they provide any.
+
+While the [PR][shada-pr] is very big, @ZyX-I has taken care of adding great test
+coverage. Great work @ZyX-I!
+
+### Wrapping the event loop layer
+
+It's no secret that libuv is the event loop library used by Neovim, and it is
+what makes it possible for us to implement features that require asynchronous
+communication (not initiated by the user) with the editor with ease.
+Unfortunately due to how Neovim code is currently organized, integrating libuv
+was not a trivial task.
+
+The basic idea is that Neovim receives arbitrary events when it is polling for
+user input, but these events can't be processed immediately because Neovim can
+be in a state that simply can't handle arbitrary actions. So if Neovim receives
+an event while checking user input, it will put the event in a queue for later
+processing.
+
+One example is illustrated in the following scenario: Neovim checks if the user
+typed ctrl+c while the regexp engine is executing, but it can't process the
+event as it is received because it may want to execute vimscript that calls the
+regexp engine again, and the engine is not reentrant since it relies heavily on
+global variables. So it has to postpone the event for when it's safe, and
+determining when it's safe to process events is itself another problem.
+
+Another complication of integrating with libuv is that sometimes Neovim must
+only process events from a certain source. For example, while Neovim is sending
+a msgpack-rpc call, it should only process events that come from:
+
+- the user (eg: ctrl+c to interrupt the call)
+- the file descriptor that received the msgpack-rpc call (which can be from a child
+  process stdio from a socket)
+
+To allow this kind of selective event processing, Neovim must maintain multiple
+queues that integrate with each other, and the logic to do this is very
+repetitive. In one of my [latest PRs][event-loop-pr], some libuv "classes" were
+wrapped in a way that makes managing these queues much easier.
+
+### jemalloc
+
+[jemalloc][jemalloc], a high performance general purpose memory allocator, is
+now used by default. Since Neovim makes heavy use of dynamic queues (see above)
```

Comment:
For some reason I thought that PR only added an option for turning jemalloc off without touching the default. I will fix it.